lib/launcher.rb

```diff
@@ -149,34 +177,122 @@ def write_pid
       retry
     end
 
-    def webrick_server(app, addresses, port)
+    def add_puma_server_callback(sd_notify)
+      events = ::Puma::Events.new(::Proxy::LogBuffer::Decorator.instance, ::Proxy::LogBuffer::Decorator.instance)
+      events.register(:state) do |status|
+        if status == :running
+          sd_notify.ready_all { logger.info("Smart proxy has finished launching #{sd_notify.total} puma instances, ready to serve requests.") }
+          logger.debug("Finished launching a puma instance, #{sd_notify.pending} instances to go...") if sd_notify.pending > 0
+        end
+      end
+      events
+    end
+
+    def add_puma_server(app, address, port, conn_type, sd_notify)
+      address = "[#{address}]" if address.include?(':')
+      logger.debug "Launching Puma listener at #{address} port #{port}"
+      if conn_type == :ssl
+        require 'cgi'
+        query_list = app[:SSLArgs].to_a.map do |x|
+          "#{CGI::escape(x[0].to_s)}=#{CGI::escape(x[1])}"
+        end
+        host = "ssl://#{address}:#{port}/?#{query_list.join('&')}"
+      else
+        host = address
+      end
+      logger.debug "Host URL: #{host}"
+      # the following lines are from lib/rack/handler/puma.rb#run
+      options = {Verbose: true, Port: port, Host: host}
+      conf = Rack::Handler::Puma.config(app[:app], options)
+      # install callback to notify systemd
+      events = add_puma_server_callback(sd_notify)
+      launcher = ::Puma::Launcher.new(conf, :events => events)
+      @servers << launcher
+      launcher.run
+    end
+
+    def add_webrick_server_callback(app, sd_notify)
+      app[:StartCallback] = lambda do
+        sd_notify.ready_all { logger.info("Smart proxy has finished launching #{sd_notify.total} webrick instances.\nSending NOTIFY_SOCKET, ready to serve requests.") }
+        logger.debug("Finished launching a webrick instance, #{sd_notify.pending} instances to go...") if sd_notify.pending > 0
+      end
+    end
+
+    def add_webrick_server(app, addresses, port, sd_notify)
+      # install callback to notify systemd
+      add_webrick_server_callback(app, sd_notify)
+      # initialize the server
       server = ::WEBrick::HTTPServer.new(app)
-      addresses.each {|a| server.listen(a, port)}
-      server.mount "/", Rack::Handler::WEBrick, app[:app]
+      addresses.each do |address|
+        logger.debug "Launching Webrick listener at #{address} port #{port}"
```

Comment:
Not needed, but the systemd notification can be used to also log this information making `systemctl status` more useful AFAIK. I'd consider that.

Reply:
Not sure what you mean, I like the debug output.